File: tree/BinaryTreeTraversal_BFS.java
package org.q10viking.tree;
import java.util.Queue;
import java.util.LinkedList;

//Iterative Queue base Java program to do level order traversal 
//of Binary Tree
class BNode{
	int data;
	BNode left,right;
	public BNode(int data) {
		this.left = this.right = null;
		this.data = data;
	}
}


public class BinaryTreeTraversal_BFS {
	BNode root;
	BinaryTreeTraversal_BFS(BNode root){
		this.root = root;
	}
	
	public void printLevelOrder() {
		Queue<BNode> queue = new LinkedList<BNode>();
		queue.add(root);
		while(!queue.isEmpty()) {
			BNode tmp = queue.poll();
			System.out.print(tmp.data+" ");
			if(tmp.left != null)
				queue.add(tmp.left);
			if(tmp.right != null)
				queue.add(tmp.right);
		}
	}
	
	public static void main(String[] args) {
		//create binary tree:
	    /*
		        	1
			    /       \
			   2          3
			 /   \       /  \
			4     5   null  null
		  /   \
		null null
		*/
		BNode root = new BNode(1);
		root.left = new BNode(2);
		root.left.left = new BNode(4);
		root.left.right = new BNode(5);
		root.right = new BNode(3);
		
		
		BinaryTreeTraversal_BFS tree = new BinaryTreeTraversal_BFS(root);
		System.out.println("Level order traversal of binary tree is - ");
		tree.printLevelOrder();
	}

}
/*
Level order traversal of binary tree is - 
1 2 3 4 5 
*/

File: tree/BinaryTreeTraversal_BFS.py
'''
Created on 2018年8月4日
https://www.geeksforgeeks.org/level-order-tree-traversal/
@author: Viking
'''
# python3 orgram to print level order traversal using queue


class Node:
    def __init__(self, data):
        self.data = data
        self.left = self.right = None


def printLevelOrder(root):
    if root is None:
        return

    # create an empty queue for level order traversal
    queue = []

    queue.append(root)
    while(len(queue) > 0):
        tmp = queue.pop(0)
        print(tmp.data, end=" ")
        if tmp.left is not None:
            queue.append(tmp.left)
        if tmp.right is not None:
            queue.append(tmp.right)


# Driver code to test above function
root = Node(1)
root.left = Node(2)
root.right = Node(3)
root.left.left = Node(4)
root.left.right = Node(5)

print("Level Order Traversal of binary tree is -")
printLevelOrder(root)

'''
Level Order Traversal of binary tree is -
1 2 3 4 5 
'''
